Fix crash when trying to test map in the editor

minetoblend opened the pull request:
Closes #34846 

Turns out the editor player was relying on the 1-frame delay before hitobject drawables would become alive prior to #34830.
Issue here was that with hitobjects becoming alive immediately, they already had a chance to update their state to "miss" before the player can manually set their result now.
This lead to a null pointer exception in the playfield's update look due to `JudgementResult.RawTime` being null.

Assigning a time to the result did fix the crash but the logic of marking the objectse was still broken, so tried to hook things up to update before the hitobject has a chance to update it's own state. Not 100% confident in this fix quality wise though so maybe reverting #34830 is a better option.

minetoblend committed:
Fix crash when marking previous objects as hit

bdach commented on osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs:
```diff
@@ -111,38 +116,39 @@ static IEnumerable<HitObject> enumerateHitObjects(IEnumerable<HitObject> hitObje
-            {
-                if (drawableObject.Entry == null)
-                    continue;
+                var drawableObject = DrawableRuleset.Playfield.HitObjectContainer
+                                                    .AliveObjects
+                                                    .LastOrDefault(it => it.HitObject == hitObject);
```
have not checked but am very skeptical this is going to handle nested drawable hit objects correctly

minetoblend commented on osu.Game/Screens/Edit/GameplayTest/EditorPlayer.cs:
```diff
@@ -55,12 +54,18 @@ protected override GameplayClockContainer CreateGameplayClockContainer(WorkingBe
         {
             base.LoadComplete();
 
             markPreviousObjectsHit();
```
I didn't want to make any changes to code that still seemed to be working but it looks like moving it into LoadAsyncComplete doesn't break anything so I'll do that.

bdach replied:
inline comment explaining this required

Commit:
Recursively update hit results for nested drawables

Commit:
Move `markPreviousObjectsHit` into `LoadAsyncComplete`

Commit:
Prevent npr in tests due to drawable ruleset not being available

Commit:
Revert moving `markPreviousObjectsHit` into `LoadAsyncComplete`

Running that there caused a test failure due to modifying drawables' transforms outside the update thread

Commit:
Fix mismatch in cutoff time check between `preventMissOnPreviousHitObjects` and `markPreviousObjectsHit`

minetoblend commented:
There was one more test failure due to a mismatch in which objects had a result applied and which ones were received by the score processor, which could lead to the score processor never receiving the results for some hitobjects and it's `HasCompleted` value never becoming true. 
It does pass now but not 100% sure if there aren't any other failure scenarios here.

Commit:
Move null check after loop over nested hitobjects

bdach committed:
Add one more comment

bdach committed:
Use `SingleOrDefault()` instead of `LastOrDefault()`

`LastOrDefault()` is arbitrary, and I hope this doesn't matter for
anything, because if it does, then that's utterly *horrifying*.

bdach reviewed (dismissed):
horrible hack but the previous code was already a hack so I'm not going to split hairs

peppy reviewed (changes requested):
Just a minor one.

bdach committed:
Use better guard (and reword subsequent comment)

Co-authored-by: Dean Herbert <pe@ppy.sh>
